File: 2022/julia/day2.jl
succ(n) = n == 3 ? 1 : n + 1
prev(n) = n == 1 ? 3 : n - 1
tieing(n) = n
winning(n) = succ(n)
losing(n) = prev(n)

function score(round)
    get_score(o, m) = begin
        succ(o) == m && return 6
        o == m && return 3
    end
    op_shape, my_shape = round
    get_score(op_shape, my_shape) + my_shape
end

function ints(line)
    toint = Dict(
        "X" => 1,
        "Y" => 2,
        "Z" => 3,
        "A" => 1,
        "B" => 2,
        "C" => 3,
    )
    map(l -> toint[l], split(line))
end

function strategy(round)
    op_shape, my_shape = round
    strategies = Dict(
        1 => losing(op_shape),
        2 => tieing(op_shape),
        3 => winning(op_shape),
    )
    (op_shape, strategies[my_shape])
end

function sol1(lines)
    sum(score.(ints.(lines)))
end

function sol2(lines)
    sum(score.(strategy.(ints.(lines))))
end


readlines("../input/2.txt") |> sol1 |> println
readlines("../input/2.txt") |> sol2 |> println

File: 2022/haskell/src/Day2.hs
module Day2 where

import Data.Text as T

ints "A" = 1
ints "B" = 2
ints "C" = 3
ints "X" = 1
ints "Y" = 2
ints "Z" = 3

winning 3 = 1
winning n = succ n
losing 1 = 3
losing n = pred n

parseRounds :: T.Text -> [(Int, Int)]
parseRounds contents =
  (\(op:me:[]) -> (ints op, ints me)) <$> splits
  where
    splits = T.splitOn " " <$> T.lines contents

roundScore (op, me)
  | op == me = 3
  | winning op == me = 6
  | otherwise = 0

shapeScore (_, me) = me

score = fmap (\r -> roundScore r + shapeScore r)

strategy (op, me)
  | me == 1 = (op, losing op)
  | me == 2 = (op, op)
  | me == 3 = (op, winning op)

sol1 = sum . score . parseRounds
sol2 = sum . score . fmap strategy . parseRounds
sol input = putStrLn $ mconcat [
    show $ sol1 input,
    " ",
    show $ sol2 input
  ]
